File: server/models/product.model.js
const mongoose = require("mongoose");

const SizeQuantitySchema = mongoose.Schema({
  size: {
    type: String,
    required: true,
  },
  quantity: {
    type: Number,
    required: true,
  },
});

const ProductSchema = mongoose.Schema(
  {
    productImage: {
      type: [String],
      required: true,
    },

    productName: {
      type: String,
      required: true,
    },

    price: {
      type: Number,
      required: true,
    },

    productDescription: {
      type: String,
      required: true,
    },

    categories: {
      type: [String],
      required: true,
    },

    sizeQuantities: {
      type: [SizeQuantitySchema],
      required: true,
    },

    status: {
      type: String,
      default: "live",
      enum: ["live", "sold out", "delisted"],
    },

    rating: {
      type: Number,
      default: 0.0,
    },

    discount: {
      type: Number,
      required: false,
      default: 0,
    },

    sellerName: {
      type: String,
      required: true,
    },

    wishedByUser: {
      type: [mongoose.Schema.Types.ObjectId],
      ref: "User",
      default: [],
    },

    sellerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
  },
  {
    timestamps: true,
  }
);

const Product = mongoose.model("Product", ProductSchema);

module.exports = Product;

File: eukay/lib/pages/dashboard/mappers/product_model.dart
import 'package:eukay/pages/dashboard/mappers/review_model.dart';

class ProductModel {
  final String id;
  final List<String> productImage;
  final String productName;
  final double price;
  final String productDescription;
  final List<String> categories;
  final List<SizeQuantity> sizeQuantities;
  final String status;
  final double rating;
  final double discount;
  final String sellerName;
  final String sellerId;
  final List<String> wishedByUser;
  final List<ReviewModel> reviews;

  const ProductModel({
    required this.id,
    required this.productImage,
    required this.sellerId,
    required this.productName,
    required this.price,
    required this.productDescription,
    required this.discount,
    required this.categories,
    required this.sizeQuantities,
    required this.status,
    required this.rating,
    required this.reviews,
    required this.sellerName,
    required this.wishedByUser,
  });

  factory ProductModel.fromJson(Map<String, dynamic> json) {
    return ProductModel(
      id: json["_id"],
      sellerId: json["sellerId"],
      productImage: List<String>.from(json["productImage"] ?? []),
      productName: json["productName"],
      price: (json["price"] is int)
          ? (json["price"] as int).toDouble()
          : (json["price"] as double? ?? 0.0),
      productDescription: json["productDescription"],
      categories: List<String>.from(json["categories"] ?? []),
      sizeQuantities: (json["sizeQuantities"] is List)
          ? List<SizeQuantity>.from(json["sizeQuantities"]
              .map((sizeJson) => SizeQuantity.fromJson(sizeJson)))
          : [],
      status: json["status"],
      discount: (json["discount"] is int)
          ? (json["discount"] as int).toDouble()
          : (json["discount"] as double? ?? 0.0),
      rating: (json["rating"] is int)
          ? (json["rating"] as int).toDouble()
          : (json["rating"] as double? ?? 0.0),
      reviews: (json["reviews"] is List)
          ? List<ReviewModel>.from(json["reviews"]
              .map((reviewJson) => ReviewModel.fromJson(reviewJson)))
          : [], // Handle non-list cases
      sellerName: json["sellerName"],
      wishedByUser: List<String>.from(json["wishedByUser"] ?? []),
    );
  }
}

class SizeQuantity {
  final String size;
  final int quantity;
  final String id;

  const SizeQuantity({
    required this.size,
    required this.quantity,
    required this.id,
  });

  factory SizeQuantity.fromJson(Map<String, dynamic> json) {
    return SizeQuantity(
      size: json["size"],
      quantity: json["quantity"],
      id: json["_id"],
    );
  }
}
